pipeline-to-taskrun/pkg/reconciler/pipelinetotaskrun/pipelinetotaskrun.go

```diff
@@ -0,0 +1,257 @@
+	return merr
+
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, run *v1alpha1.Run) error {
+	logger := logging.FromContext(ctx)
+
+	// confirm the run spec is valid
+	if err := validateRun(run); err != nil {
+		logger.Errorf("Run %s/%s is invalid because of %v", run.Namespace, run.Name, err)
+		run.Status.MarkRunFailed(ReasonRunFailedValidation,
+			"Run has an invalid spec: %v", err)
+		return controller.NewPermanentError(fmt.Errorf("run %s/%s is invalid because of %v", run.Namespace, run.Name, err))
+	}
+
+	// fetch the taskrun and, if present, update the run status
+	tr, err := getTaskRunIfExists(r.taskRunLister, run.Namespace, run.Name)
+	if err != nil {
+		logger.Errorf("Run %s/%s got an error fetching taskRun: %v", run.Namespace, run.Name, err)
+		return fmt.Errorf("couldn't fetch taskrun %v", err)
+	}
+	if tr != nil {
+		logger.Infof("Found a TaskRun object %s", tr.Name)
+		return updateRunStatus(ctx, run, tr)
+	}
+
+	// get the pipeline that we're going to be running in a taskrun
+	pSpec, err := getPipelineSpec(ctx, r.pipelineClientSet.TektonV1beta1(), run.Namespace, run.Spec.Ref.Name)
+	if err != nil {
+		run.Status.MarkRunFailed(ReasonRunFailedValidation,
+			"Pipeline couldn't be fetched - %v", err)
+		return controller.NewPermanentError(fmt.Errorf("run %s/%s is invalid because of %v", run.Namespace, run.Name, err))
+		return fmt.Errorf("couldn't fetch pipeline spec %s: %v", run.Spec.Ref.Name, err)
+	}
+	if err := validatePipelineSpec(pSpec); err != nil {
+		run.Status.MarkRunFailed(ReasonRunFailedValidation,
+			"Pipeline is invalid - %v", err)
+		return fmt.Errorf("pipeline spec for %s is invalid: %v", run.Spec.Ref.Name, err)
+	}
+
+	// get all the tasks we need to run this pipeline
+	taskSpecs, err := getTaskSpecs(ctx, r.pipelineClientSet.TektonV1beta1(), pSpec, run.Namespace)
+	if err != nil {
+		run.Status.MarkRunFailed(ReasonRunFailedValidation,
+			"Not all of the pipeline's tasks could be fetched - %v", err)
+		return fmt.Errorf("couldn't fetch pipeline's tasks: %v", err)
+	}
+	if err := validateTaskSpecs(taskSpecs); err != nil {
+		run.Status.MarkRunFailed(ReasonRunFailedValidation,
+			"Not all tasks are valid - %v", err)
+		return fmt.Errorf("pipeline's tasks are invalid: %v", err)
+	}
+
+	tr, err = getMergedTaskRun(run, pSpec, taskSpecs)
+	if err != nil {
+		run.Status.MarkRunFailed(ReasonRunFailedValidation,
+			"Could not create TaskRun for the pipeline - %v", err)
+		return fmt.Errorf("couldn't create taskrun for pipeline %s: %v", run.Spec.Ref.Name, err)
+	}
+
+	logger.Infof("Creating a new TaskRun object %s", tr.Name)
+	if _, err := r.pipelineClientSet.TektonV1beta1().TaskRuns(run.Namespace).Create(ctx, tr, metav1.CreateOptions{}); err != nil {
+		logger.Errorf("Run %s/%s got an error creating TaskRun - %v", run.Namespace, run.Name, err)
+		run.Status.MarkRunFailed(ReasonRunFailedCreatingPipelineRun,
+			"Run got an error creating pipelineRun - %v", err)
+		return err
+	}
```

Comment:
following this section is a bit confusing i.e. it's not clear when taskrun is created 

e.g. we have an error "couldn't create taskrun for pipeline", then later we're "creating a new taskrun object"

may help if the first part used `merge` and second part part used `create`

Reply:
or you could add comments like in the other sections above, they were really helpful - thanks!

Reply:
both great suggestions! ive made a couple updates but lemme know if it's not enough of a change and i could still make it more clear